internal/rotate/file_writer.go

```diff
@@ -0,0 +1,176 @@
+	if interval == 0 && maxSizeInBytes <= 0 {
+		// No rotation needed so a basic io.Writer will do the trick
+		return openFile(filename)
+	}
+
+	w := &FileWriter{filename: filename, interval: interval, maxSizeInBytes: maxSizeInBytes,
+		maxArchives: maxArchives, filenameRotationTemplate: getFilenameRotationTemplate(filename)}
+
+	if err := w.openCurrent(true); err != nil {
+		return nil, err
+	}
+
+	return w, nil
+}
+
+func openFile(filename string) (*os.File, error) {
+	return os.OpenFile(filename, os.O_RDWR|os.O_CREATE|os.O_APPEND, FilePerm)
+}
+
+func getFilenameRotationTemplate(filename string) string {
+	// Extract the file extension
+	fileExt := filepath.Ext(filename)
+	// Remove the file extension from the filename (if any)
+	filenameWithoutExtension := strings.TrimSuffix(filename, fileExt)
+	return filenameWithoutExtension + "-%s-%s" + fileExt
+}
+
+// Write writes p to the current file, then checks to see if
+// rotation is necessary.
+func (w *FileWriter) Write(p []byte) (n int, err error) {
+	w.Lock()
+	defer w.Unlock()
+	if n, err = w.current.Write(p); err != nil {
+		return 0, err
+	}
+	w.bytesWritten += int64(n)
+
+	if err = w.rotateIfNeeded(); err != nil {
+		return 0, err
+	}
+
+	return n, nil
+}
+
+// Close closes the current file.  Writer is unusable after this
+// is called.
+func (w *FileWriter) Close() (err error) {
+	w.Lock()
+	defer w.Unlock()
+
+	// Rotate before closing
+	if err = w.rotate(false); err != nil {
+		return err
+	}
+
+	if err = w.current.Close(); err != nil {
+		return err
+	}
+	w.current = nil
+	return nil
+}
+
+func (w *FileWriter) openCurrent(firstRun bool) (err error) {
+	w.current, err = openFile(w.filename)
+	w.expireTime = time.Now().Add(w.interval)
+	w.bytesWritten = 0
+
+	if err != nil {
+		return err
+	}
+	if !firstRun {
+		return nil
+	}
+
+	// Goal here is to rotate old pre-existing files. For that we use fileInfo.ModTime, instead of time.Now(), only when the FileWriter is created (firstRun)
+	// Example: telegraf is restarted every 23 hours and the rotation interval is set to 24 hours. With time.now() as a reference we'd never rotate the file.
+	// It's only necessary to use modtime when the filewriter is created, otherwise we assume that we've been continuously running, so time.now is fine.
+	if fileInfo, err := w.current.Stat(); err == nil {
+		w.expireTime = fileInfo.ModTime().Add(w.interval)
+	}
+	return nil
+}
+
+func (w *FileWriter) rotateIfNeeded() error {
+	if (w.interval > 0 && time.Now().After(w.expireTime)) ||
+		(w.maxSizeInBytes > 0 && w.bytesWritten >= w.maxSizeInBytes) {
+		return w.rotate(true)
+	}
+	return nil
+}
+
+func (w *FileWriter) rotate(createNewFile bool) (err error) {
+	if err = w.current.Close(); err != nil {
+		return err
+	}
+	now := time.Now()
+	// Use year-month-date for readability, unix time to make archive names unique
+	rotatedFilename := fmt.Sprintf(w.filenameRotationTemplate, now.Format(DateFormat), strconv.FormatInt(now.UnixNano(), 10))
```

Comment:
File name can't contain `:` :(
We could do `telegraf.2019-04-30T163936.log` instead ?

Reply:
Forgot about Windows file name restrictions, I think without the colons it is quite a bit harder to understand, so maybe just like you had it before but with time.Unix() instead of UnixNano?